File: CS/SchedulerGSync/SchedulerGSync/MainForm.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Globalization;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Forms;
using DevExpress.XtraScheduler;
using Google.Apis.Auth.OAuth2;
using Google.Apis.Calendar.v3;
using Google.Apis.Services;
using Google.Apis.Util.Store;
using Microsoft.EntityFrameworkCore;
using SchedulerGSync.DAL;
using System.Linq;
using DevExpress.XtraEditors;

namespace SchedulerGSync {
    public partial class MainForm : XtraForm {
        SchedulerGSyncDbContext dbContext;
        readonly string[] scopes = { CalendarService.Scope.Calendar };
        UserCredential credential;
        CalendarService service;
        string defaultLabelKey = String.Empty;

        public MainForm() {
            InitializeComponent();
            Text = $"{nameof(SchedulerGSync)}";
            this.dbContext = new SchedulerGSyncDbContext();            
        }

        void InitScheduler() {
            this.schedulerControl1.DayView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.DayView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.WorkWeekView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.WorkWeekView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.FullWeekView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.FullWeekView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.MonthView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never;
            this.schedulerControl1.AgendaView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never;

            this.dbContext.Database.EnsureCreated();
            this.dbContext.AppointmentObjects.Load();
            BindingList<AppointmentObject> bindingList = this.dbContext.AppointmentObjects.Local.ToBindingList();
            BindDataToStorage(bindingList);
            this.dxGoogleCalendarSync1.EventValuesRequested += OnEventValuesRequested;
            this.dxGoogleCalendarSync1.AppointmentValuesRequested += OnAppointmentValuesRequested;
            this.dxGoogleCalendarSync1.CompareEventAndAppointment += OnCompareEventAndAppointment;
        }

        void OnCompareEventAndAppointment(object sender, DevExpress.XtraScheduler.GoogleCalendar.CompareEventAndAppointmentEventArgs e) {
            if (!e.IsEqual)
                return;
            var eventColorId = e.Event.ColorId ?? this.defaultLabelKey;
            e.IsEqual = eventColorId == (string)e.Appointment.LabelKey;
        }

        void OnAppointmentValuesRequested(object sender, DevExpress.XtraScheduler.GoogleCalendar.ObjectValuesRequestedEventArgs e) {
            if ((string)e.Appointment.LabelKey == this.defaultLabelKey)
                e.Event.ColorId = null;
            else
                e.Event.ColorId = (string)e.Appointment.LabelKey;
        }

        void OnEventValuesRequested(object sender, DevExpress.XtraScheduler.GoogleCalendar.ObjectValuesRequestedEventArgs e) {
            e.Appointment.LabelKey = e.Event.ColorId ?? this.defaultLabelKey;
        }

        void OnBbiSyncItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            this.dxGoogleCalendarSync1.CalendarId = ApplicationConfig.Instance.CalendarId;
            this.dxGoogleCalendarSync1.CalendarService = this.service;
            this.dxGoogleCalendarSync1.Synchronize();
        }

        async protected override void OnLoad(EventArgs e) {
            base.OnLoad(e);
            try {
                this.credential = await AuthorizeToGoogle();
                this.service = new CalendarService(new BaseClientService.Initializer() {
                    HttpClientInitializer = this.credential,
                    ApplicationName = nameof(SchedulerGSync)
                });
                await UpdateCalendarRelatedData();
                InitScheduler();
            } catch (Exception ex) {
                MessageBox.Show(ex.Message);
            }
        }
        
        async Task UpdateLabels() {
            this.schedulerDataStorage1.Labels.Clear();
            this.schedulerDataStorage1.Statuses.Clear();
            string calendarId = ApplicationConfig.Instance.CalendarId;
            this.defaultLabelKey = String.Empty;
            if (String.IsNullOrEmpty(calendarId))
                return;
            var colors = await this.service.Colors.Get().ExecuteAsync();
            var calendar = await this.service.CalendarList.Get(calendarId).ExecuteAsync();
            Color defaultColor = ColorHelper.FromString(colors.Calendar[calendar.ColorId].Background);
            foreach (var colorDefinition in colors.Event__) {
                Color color = ColorHelper.FromString(colorDefinition.Value.Background);
                var label = this.schedulerDataStorage1.Labels.Items.CreateNewLabel(colorDefinition.Key, $"Color {colorDefinition.Key}", $"Color {colorDefinition.Key}", color);
                this.schedulerDataStorage1.Labels.Items.Add(label);
                if (color == defaultColor)
                    this.defaultLabelKey = colorDefinition.Key;
            }
            if (String.IsNullOrEmpty(defaultLabelKey)) {
                this.defaultLabelKey = "Calendar";
                var defaultLabel = this.schedulerDataStorage1.Labels.Items.CreateNewLabel(defaultLabelKey, $"Default", $"Default", defaultColor);
                this.schedulerDataStorage1.Labels.Items.Add(defaultLabel);
            }
        }
                
        async Task<UserCredential> AuthorizeToGoogle() {
            using (FileStream stream = new FileStream("client_secret.json", FileMode.Open, FileAccess.Read)) {
                string credPath = Environment.GetFolderPath(
                    Environment.SpecialFolder.Personal);
                credPath = Path.Combine(credPath, $".credentials/{nameof(SchedulerGSync)}.json");

                return await GoogleWebAuthorizationBroker.AuthorizeAsync(
                    GoogleClientSecrets.Load(stream).Secrets,
                    this.scopes,
                    "user",
                    CancellationToken.None,
                    new FileDataStore(credPath, true));
            }
        }

        protected override void OnFormClosing(FormClosingEventArgs e) {
            base.OnFormClosing(e);
            this.dbContext.SaveChanges();
            ApplicationConfig.Instance.Save();
        }

        void BindDataToStorage(BindingList<AppointmentObject> bindingList) {
            var mappings = this.schedulerDataStorage1.Appointments.Mappings;
            mappings.Start = "Start";
            mappings.End = "End";
            mappings.Subject = "Subject";
            mappings.Type = "AppointmentType";
            mappings.RecurrenceInfo = "RecurrenceInfo";
            mappings.ReminderInfo = "ReminderInfo";
            mappings.Location = "Location";
            mappings.Label = "Label";
            this.schedulerDataStorage1.Appointments.CustomFieldMappings.Add(new AppointmentCustomFieldMapping("gId", "GId"));
            this.schedulerDataStorage1.Appointments.CustomFieldMappings.Add(new AppointmentCustomFieldMapping("etag", "ETag"));
            this.schedulerDataStorage1.Appointments.DataSource = bindingList;
        }

        async void OnBbiOptionsItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            SyncOptionsForm form = new SyncOptionsForm(this.service);
            form.CalendarId = ApplicationConfig.Instance.CalendarId;
            if (form.ShowDialog(this) == DialogResult.OK) {
                if (ApplicationConfig.Instance.CalendarId != form.CalendarId) {
                    ClearSchedulerData();
                    ApplicationConfig.Instance.CalendarId = form.CalendarId;
                    await UpdateCalendarRelatedData();
                }
            }
        }

        async Task UpdateCalendarRelatedData() {
            await UpdateLabels();
            await UpdateCaption();
        }

        void ClearSchedulerData() {
            this.schedulerDataStorage1.Appointments.DataSource = null;

            this.dbContext.Dispose();
            File.Delete(SchedulerGSyncDbContext.DBFileName);
            this.dbContext = new SchedulerGSyncDbContext();
            this.dbContext.Database.EnsureCreated();
            this.dbContext.AppointmentObjects.Load();

            this.schedulerDataStorage1.Appointments.DataSource = this.dbContext.AppointmentObjects.Local.ToBindingList();
            this.dxGoogleCalendarSync1.Save();
        }

        async Task UpdateCaption() {
            string calendarId = ApplicationConfig.Instance.CalendarId;
            if (String.IsNullOrEmpty(calendarId))
                return;
            var calendar = await this.service.Calendars.Get(calendarId).ExecuteAsync();
            Text = $"{nameof(SchedulerGSync)} - {calendar.Summary}";
        }
    }
}

File: VB/SchedulerGSync/SchedulerGSync/MainForm.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Globalization
Imports System.IO
Imports System.Threading
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler
Imports Google.Apis.Auth.OAuth2
Imports Google.Apis.Calendar.v3
Imports Google.Apis.Services
Imports Google.Apis.Util.Store
Imports Microsoft.EntityFrameworkCore
Imports SchedulerGSync.DAL
Imports System.Linq
Imports DevExpress.XtraEditors

Namespace SchedulerGSync
	Partial Public Class MainForm
		Inherits XtraForm

		Private dbContext As SchedulerGSyncDbContext
		Private ReadOnly scopes() As String = { CalendarService.Scope.Calendar }
		Private credential As UserCredential
		Private service As CalendarService
		Private defaultLabelKey As String = String.Empty

		Public Sub New()
			InitializeComponent()
			Text = $"{NameOf(SchedulerGSync)}"
			Me.dbContext = New SchedulerGSyncDbContext()
		End Sub

		Private Sub InitScheduler()
			Me.schedulerControl1.DayView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.DayView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.WorkWeekView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.WorkWeekView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.FullWeekView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.FullWeekView.AppointmentDisplayOptions.AllDayAppointmentsStatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.MonthView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never
			Me.schedulerControl1.AgendaView.AppointmentDisplayOptions.StatusDisplayType = AppointmentStatusDisplayType.Never

			Me.dbContext.Database.EnsureCreated()
			Me.dbContext.AppointmentObjects.Load()
			Dim bindingList As BindingList(Of AppointmentObject) = Me.dbContext.AppointmentObjects.Local.ToBindingList()
			BindDataToStorage(bindingList)
			AddHandler dxGoogleCalendarSync1.EventValuesRequested, AddressOf OnEventValuesRequested
			AddHandler dxGoogleCalendarSync1.AppointmentValuesRequested, AddressOf OnAppointmentValuesRequested
			AddHandler dxGoogleCalendarSync1.CompareEventAndAppointment, AddressOf OnCompareEventAndAppointment
		End Sub

		Private Sub OnCompareEventAndAppointment(ByVal sender As Object, ByVal e As DevExpress.XtraScheduler.GoogleCalendar.CompareEventAndAppointmentEventArgs)
			If Not e.IsEqual Then
				Return
			End If
			Dim eventColorId = If(e.Event.ColorId, Me.defaultLabelKey)
			e.IsEqual = eventColorId = DirectCast(e.Appointment.LabelKey, String)
		End Sub

		Private Sub OnAppointmentValuesRequested(ByVal sender As Object, ByVal e As DevExpress.XtraScheduler.GoogleCalendar.ObjectValuesRequestedEventArgs)
			If CStr(e.Appointment.LabelKey) = Me.defaultLabelKey Then
				e.Event.ColorId = Nothing
			Else
				e.Event.ColorId = CStr(e.Appointment.LabelKey)
			End If
		End Sub

		Private Sub OnEventValuesRequested(ByVal sender As Object, ByVal e As DevExpress.XtraScheduler.GoogleCalendar.ObjectValuesRequestedEventArgs)
			e.Appointment.LabelKey = If(e.Event.ColorId, Me.defaultLabelKey)
		End Sub

		Private Sub OnBbiSyncItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles bbiSync.ItemClick
			Me.dxGoogleCalendarSync1.CalendarId = ApplicationConfig.Instance.CalendarId
			Me.dxGoogleCalendarSync1.CalendarService = Me.service
			Me.dxGoogleCalendarSync1.Synchronize()
		End Sub

		Async Protected Overrides Sub OnLoad(ByVal e As EventArgs)
			MyBase.OnLoad(e)
			Try
				Me.credential = Await AuthorizeToGoogle()
				Me.service = New CalendarService(New BaseClientService.Initializer() With {
					.HttpClientInitializer = Me.credential,
					.ApplicationName = NameOf(SchedulerGSync)
				})
				Await UpdateCalendarRelatedData()
				InitScheduler()
			Catch ex As Exception
				MessageBox.Show(ex.Message)
			End Try
		End Sub

		Private Async Function UpdateLabels() As Task
			Me.schedulerDataStorage1.Labels.Clear()
			Me.schedulerDataStorage1.Statuses.Clear()
			Dim calendarId As String = ApplicationConfig.Instance.CalendarId
			Me.defaultLabelKey = String.Empty
			If String.IsNullOrEmpty(calendarId) Then
				Return
			End If
			Dim colors = Await Me.service.Colors.Get().ExecuteAsync()
			Dim calendar = Await Me.service.CalendarList.Get(calendarId).ExecuteAsync()
			Dim defaultColor As Color = ColorHelper.FromString(colors.Calendar(calendar.ColorId).Background)
			For Each colorDefinition In colors.Event__
				Dim color As Color = ColorHelper.FromString(colorDefinition.Value.Background)
				Dim label = Me.schedulerDataStorage1.Labels.Items.CreateNewLabel(colorDefinition.Key, $"Color {colorDefinition.Key}", $"Color {colorDefinition.Key}", color)
				Me.schedulerDataStorage1.Labels.Items.Add(label)
				If color = defaultColor Then
					Me.defaultLabelKey = colorDefinition.Key
				End If
			Next colorDefinition
			If String.IsNullOrEmpty(defaultLabelKey) Then
				Me.defaultLabelKey = "Calendar"
				Dim defaultLabel = Me.schedulerDataStorage1.Labels.Items.CreateNewLabel(defaultLabelKey, $"Default", $"Default", defaultColor)
				Me.schedulerDataStorage1.Labels.Items.Add(defaultLabel)
			End If
		End Function

		Private Async Function AuthorizeToGoogle() As Task(Of UserCredential)
			Using stream As New FileStream("client_secret.json", FileMode.Open, FileAccess.Read)
				Dim credPath As String = Environment.GetFolderPath(Environment.SpecialFolder.Personal)
				credPath = Path.Combine(credPath, $".credentials/{NameOf(SchedulerGSync)}.json")

				Return Await GoogleWebAuthorizationBroker.AuthorizeAsync(GoogleClientSecrets.Load(stream).Secrets, Me.scopes, "user", CancellationToken.None, New FileDataStore(credPath, True))
			End Using
		End Function

		Protected Overrides Sub OnFormClosing(ByVal e As FormClosingEventArgs)
			MyBase.OnFormClosing(e)
			Me.dbContext.SaveChanges()
			ApplicationConfig.Instance.Save()
		End Sub

		Private Sub BindDataToStorage(ByVal bindingList As BindingList(Of AppointmentObject))
			Dim mappings = Me.schedulerDataStorage1.Appointments.Mappings
			mappings.Start = "Start"
			mappings.End = "End"
			mappings.Subject = "Subject"
			mappings.Type = "AppointmentType"
			mappings.RecurrenceInfo = "RecurrenceInfo"
			mappings.ReminderInfo = "ReminderInfo"
			mappings.Location = "Location"
			mappings.Label = "Label"
			Me.schedulerDataStorage1.Appointments.CustomFieldMappings.Add(New AppointmentCustomFieldMapping("gId", "GId"))
			Me.schedulerDataStorage1.Appointments.CustomFieldMappings.Add(New AppointmentCustomFieldMapping("etag", "ETag"))
			Me.schedulerDataStorage1.Appointments.DataSource = bindingList
		End Sub

		Private Async Sub OnBbiOptionsItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs) Handles bbiOptions.ItemClick
			Dim form As New SyncOptionsForm(Me.service)
			form.CalendarId = ApplicationConfig.Instance.CalendarId
			If form.ShowDialog(Me) = DialogResult.OK Then
				If ApplicationConfig.Instance.CalendarId <> form.CalendarId Then
					ClearSchedulerData()
					ApplicationConfig.Instance.CalendarId = form.CalendarId
					Await UpdateCalendarRelatedData()
				End If
			End If
		End Sub

		Private Async Function UpdateCalendarRelatedData() As Task
			Await UpdateLabels()
			Await UpdateCaption()
		End Function

		Private Sub ClearSchedulerData()
			Me.schedulerDataStorage1.Appointments.DataSource = Nothing

			Me.dbContext.Dispose()
			File.Delete(SchedulerGSyncDbContext.DBFileName)
			Me.dbContext = New SchedulerGSyncDbContext()
			Me.dbContext.Database.EnsureCreated()
			Me.dbContext.AppointmentObjects.Load()

			Me.schedulerDataStorage1.Appointments.DataSource = Me.dbContext.AppointmentObjects.Local.ToBindingList()
			Me.dxGoogleCalendarSync1.Save()
		End Sub

		Private Async Function UpdateCaption() As Task
			Dim calendarId As String = ApplicationConfig.Instance.CalendarId
			If String.IsNullOrEmpty(calendarId) Then
				Return
			End If
			Dim calendar = Await Me.service.Calendars.Get(calendarId).ExecuteAsync()
			Text = $"{NameOf(SchedulerGSync)} - {calendar.Summary}"
		End Function
	End Class
End Namespace
